### ISSUE TITLE
disallow `name` in Options API component

### ISSUE THREAD

**Issue by privatenumber** [vuejs/eslint-plugin-vue#2087]
**Please describe what the rule should do:**

Disallow `name` from being set on Option API style components.

The `name` property should generally be the file name, and since vue-loader defaults the component name with the file name, it's not necessary to set manually.

This rule will be helpful in discouraging the practice and removing unused strings in the codebase for a smaller bundle size.

FYI In Vue 3, there's no way to set the component name; it's just the file name.

**What category should the rule belong to?**

[x] Enforces code style (layout)
[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```vue
<script>
export default {
  name: 'ComponentName'
})
</script>
```

However, I think recursive components should be fine (unless vue-loader is able to resolve the self-recursion by file name):
```vue
<template>
	<div>
		<component-name />
	</div>
</template>
<script>
export default {
  name: 'ComponentName'
})
</script>
```

**Additional context**

Perhaps this rule can be merged with https://eslint.vuejs.org/rules/require-name-property.html

**Comment by ota-meshi** [comment]
Thank you for suggesting a new rule.

> Perhaps this rule can be merged with https://eslint.vuejs.org/rules/require-name-property.html

I think it makes sense to deprecate the require-name-property rule and add a new rule to optionally disallow/allow the name option.

If you want to disallow the name property early on, you may be able to use the no-restricted-component-options rule. (Note, however, that recursive component uses are not checked.)
https://eslint.vuejs.org/rules/no-restricted-component-options.html
